addons/stock/models/stock_move.py

```diff
@@ -701,6 +701,7 @@ def write(self, vals):
         if move_to_recompute_state:
             move_to_recompute_state._recompute_state()
         if move_to_check_location:
+            move_to_check_location.warehouse_id = self.env['stock.location'].browse(vals['location_id']).warehouse_id
```

Comment:
Add a fallback on location_dest_id. In a receipt supplier is not in a warehouse but we want to use the warehouse of destination